awx/main/models/credential/__init__.py

```diff
@@ -430,6 +430,16 @@ def unique_dict(cred_qs):
             ret[cred.unique_hash()] = cred
         return ret
 
+    def get_input(self, field_name, **kwargs):
+        if field_name in self.credential_type.secret_fields:
+            return decrypt_field(self, field_name)
+        if 'default' in kwargs:
+            return getattr(self, field_name, kwargs['default'])
```

Comment:
This default pattern confuses me, because the field itself may have a default. So if the default is defined in both places, which wins?

Reply:
Yeah : \. Some of the access patterns we're replacing involve `getattr`, both with _and_ without a default. When getattr is used without a default, it raises an `AttributeError`. These lines were added to keep that behavior consistent. If you know of a better way I'd be more than happy to take a look.

Regarding your question: In the scenario you described, the field default value would win.

Reply:
> The default keyword is a bit of a hack, but it lets us have similar behavior from the perspective of the caller while helping the get_input implementation decide whether it needs to call an external backend.

I think there's just more context I need about the feature plan. Given the line you shared there, I don't see how we could possibly avoid a call to the external backend (if the credential uses one). You could cache the values, but I don't think that's what you're talking about.

Reply:
@AlanCoding 
>   I don't see how we could possibly avoid a call to the external backend


The credential will know which of its fields have been related to an external backend by maintaining a list of those fields. If a request is made for a field that is undefined and not related to a secret backend, the `default=` helps the implementation know whether or not to raise an AttributeError or return the default value - all of this can happen without talking to the secret backend.

Does that make sense?